### PR TITLE
Supports for DETR training mode in process.

### PR DESCRIPTION
**Pull request by v-qjqs**
This PR implements DETR training mode in process.

### CODE REVIEW

File: mmdet/models/dense_heads/transformer_head.py [hunk 1]
```diff
@@ -245,25 +246,337 @@ def forward_single(self, x, img_metas):
...
+        # loss from the last decoder layer
+        loss_dict['loss_cls'] = losses_cls[-1]
+        loss_dict['loss_bbox'] = losses_bbox[-1]
+        loss_dict['loss_iou'] = losses_iou[-1]
+        # loss from other decoder layers
+        num_dec_layer = 0
+        for loss_cls_i, loss_bbox_i, loss_iou_i in zip(losses_cls[:-1],
+                                                       losses_bbox[:-1],
+                                                       losses_iou[:-1]):
+            loss_dict['loss_cls_{}'.format(num_dec_layer)] = loss_cls_i
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:316]: Use f-string,

File: mmdet/models/dense_heads/transformer_head.py [hunk 2]
```diff
@@ -245,25 +246,337 @@ def forward_single(self, x, img_metas):
...
+        labels = torch.cat(labels_list, 0)
+        label_weights = torch.cat(label_weights_list, 0)
+        bbox_targets = torch.cat(bbox_targets_list, 0)
+        bbox_weights = torch.cat(bbox_weights_list, 0)
+
+        # construct factors used for rescale bboxes
+        factors = []
+        for img_meta, bbox_pred in zip(img_metas, bbox_preds):
+            img_h, img_w, _ = img_meta['img_shape']
+            factor = torch.Tensor([img_w, img_h, img_w,
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:367]: bbox_pred.new_tensor(xxxxx)

File: mmdet/models/dense_heads/transformer_head.py [hunk 3]
```diff
@@ -245,25 +246,337 @@ def forward_single(self, x, img_metas):
...
+
+        # construct weighted avg_factor to match with the official DETR repo
+        cls_avg_factor = num_total_pos * 1.0 + \
+            num_total_neg * self.bg_cls_weight
+        loss_cls = self.loss_cls(
+            cls_scores, labels, label_weights, avg_factor=cls_avg_factor)
+
+        # Compute the average number of gt boxes accross all gpus, for
+        # normalization purposes
+        num_total_pos = torch.as_tensor([num_total_pos],
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:385]: Use existing `reduce_mean` function in the code.

File: mmdet/models/dense_heads/transformer_head.py [hunk 4]
```diff
@@ -245,25 +246,337 @@ def forward_single(self, x, img_metas):
...
+                                    dtype=torch.long)
+        labels[pos_inds] = gt_labels[sampling_result.pos_assigned_gt_inds]
+        label_weights = gt_bboxes.new_ones(num_bboxes)
+
+        # bbox targets
+        bbox_targets = torch.zeros_like(bbox_pred)
+        bbox_weights = torch.zeros_like(bbox_pred)
+        bbox_weights[pos_inds] = 1.0
+        img_h, img_w, _ = img_meta['img_shape']
+        factor = torch.Tensor([img_w, img_h, img_w,
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:519]: bbox_pred.new_tensor(xxxxxx)

File: mmdet/models/detectors/single_stage.py [hunk 5]
```diff
@@ -89,6 +89,13 @@ def forward_train(self,
         Returns:
             dict[str, Tensor]: A dictionary of loss components.
         """
+        # NOTE the batched image size information may be useful, e.g.
```
Comment (ZwwWayne) [inline mmdet/models/detectors/single_stage.py:92]: Are these modification duplicate? Or should we move it into base detector.

File: tools/train.py [hunk 6]
```diff
@@ -143,6 +143,11 @@ def main():
     if args.seed is not None:
         logger.info(f'Set random seed to {args.seed}, '
                     f'deterministic: {args.deterministic}')
+
+        # import torch.distributed as dist
```
Comment (ZwwWayne) [inline tools/train.py:147]: clean unnecessary modification.

File: mmdet/models/dense_heads/transformer_head.py [hunk 7]
```diff
@@ -245,25 +248,332 @@ def forward_single(self, x, img_metas):
...
+                for all images. Shape [bs, num_query, cls_out_channels].
+            bbox_preds (Tensor): Sigmoid outputs from a single decoder layer
+                for all images, with normalized coordinate (cx, cy, w, h) and
+                shape [bs, num_query, 4].
+            gt_bboxes_list (list[Tensor]): Ground truth bboxes for each image
+                with shape (num_gts, 4) in [tl_x, tl_y, br_x, br_y] format.
+            gt_labels_list (list[Tensor]): Ground truth class indices for each
+                image with shape (num_gts, ).
+            img_metas (list[dict]): List of image meta information.
+            gt_bboxes_ignore_list (None | list[Tensor], optional): Bounding
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:345]: (list[Tensor], optional): xxxx. Defaults: None.

File: mmdet/datasets/pipelines/transforms.py [hunk 8]
```diff
@@ -80,6 +81,7 @@ def __init__(self,
         self.multiscale_mode = multiscale_mode
         self.ratio_range = ratio_range
         self.keep_ratio = keep_ratio
+        self.call_repeat = call_repeat
```
Comment (ZwwWayne) [inline mmdet/datasets/pipelines/transforms.py:84]: call_repeat -> override

File: mmdet/models/dense_heads/transformer_head.py [hunk 9]
```diff
@@ -245,25 +248,332 @@ def forward_single(self, x, img_metas):
...
+                                           gt_bboxes_list, gt_labels_list,
+                                           img_metas, gt_bboxes_ignore_list)
+        (labels_list, label_weights_list, bbox_targets_list, bbox_weights_list,
+         num_total_pos, num_total_neg) = cls_reg_targets
+        labels = torch.cat(labels_list, 0)
+        label_weights = torch.cat(label_weights_list, 0)
+        bbox_targets = torch.cat(bbox_targets_list, 0)
+        bbox_weights = torch.cat(bbox_weights_list, 0)
+
+        # construct factors used for rescale bboxes
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:365]: Put lines 366-373 to the places that directly above line 395 (the place it will be used).

File: mmdet/models/dense_heads/transformer_head.py [hunk 10]
```diff
@@ -245,25 +248,332 @@ def forward_single(self, x, img_metas):
...
+        assign_result = self.assigner.assign(bbox_pred, cls_score, gt_bboxes,
+                                             gt_labels, img_meta,
+                                             gt_bboxes_ignore)
+        sampling_result = self.sampler.sample(assign_result, bbox_pred,
+                                              gt_bboxes)
+        pos_inds = sampling_result.pos_inds
+        neg_inds = sampling_result.neg_inds
+
+        # label targets
+        labels = gt_bboxes.new_full((num_bboxes, ),
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:505]: Can this part of code can handle the case when `len(pos_inds) == 0`? If not, we should support that as those in other heads.

Comment (v-qjqs) [reply mmdet/models/dense_heads/transformer_head.py:505]: Hi, it can handle case when there is no gt boxes. In this case, there exists only cls loss, iou loss and bbox loss will be zero.

File: mmdet/models/dense_heads/transformer_head.py [hunk 11]
```diff
@@ -245,25 +248,332 @@ def forward_single(self, x, img_metas):
...
+        num_total_pos = loss_cls.new_tensor([num_total_pos])
+        num_total_pos = torch.clamp(reduce_mean(num_total_pos), min=1).item()
+
+        # regression L1 loss
+        bbox_preds = bbox_preds.reshape(-1, 4)
+        loss_bbox = self.loss_bbox(
+            bbox_preds, bbox_targets, bbox_weights, avg_factor=num_total_pos)
+
+        # regression iou loss, defaultly giou loss
+        bboxes = bbox_cxcywh_to_xyxy(bbox_preds) * factors
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:395]: Leave a comments here to indicate why factors is necessary here. For example, 'DETR regress the relative position of boxes in the image, thus, the learning target is normalized by the image size'.

File: mmdet/models/dense_heads/transformer_head.py [hunk 12]
```diff
@@ -245,25 +248,332 @@ def forward_single(self, x, img_metas):
...
+        label_weights = gt_bboxes.new_ones(num_bboxes)
+
+        # bbox targets
+        bbox_targets = torch.zeros_like(bbox_pred)
+        bbox_weights = torch.zeros_like(bbox_pred)
+        bbox_weights[pos_inds] = 1.0
+        img_h, img_w, _ = img_meta['img_shape']
+        factor = bbox_pred.new_tensor([img_w, img_h, img_w,
+                                       img_h]).unsqueeze(0)
+        pos_gt_bboxes_normalized = sampling_result.pos_gt_bboxes / factor
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:518]: Leave a comment here to indicate why `factors` is necessary here. For example, 'DETR regress the relative position of boxes in the image, thus, the learning target is normalized by the image size'.

File: mmdet/models/dense_heads/transformer_head.py [hunk 13]
```diff
@@ -163,15 +167,14 @@ def _init_layers(self):
...
             add_residual=False)
         self.fc_reg = Linear(self.embed_dims, 4)
         self.query_embedding = nn.Embedding(self.num_query, self.embed_dims)
 
     def init_weights(self, distribution='uniform'):
         """Initialize weights of the transformer head."""
-        for m in self.modules():
-            if hasattr(m, 'weight') and m.weight.dim() > 1:
-                xavier_init(m, distribution=distribution)
+        self.transformer.init_weights()
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:177]: Add comments here, indicating the initialization is important.

File: mmdet/models/dense_heads/transformer_head.py [hunk 14]
```diff
@@ -245,25 +248,339 @@ def forward_single(self, x, img_metas):
...
+    def loss(self,
+             all_cls_scores_list,
+             all_bbox_preds_list,
+             gt_bboxes_list,
+             gt_labels_list,
+             img_metas,
+             gt_bboxes_ignore=None):
+        """"Loss function.
+
+        Defaultly only outputs from the last feature level is used for
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:267]: Only outputs from the last feature level are used for computing losses by default.

File: mmdet/models/dense_heads/transformer_head.py [hunk 15]
```diff
@@ -245,25 +248,339 @@ def forward_single(self, x, img_metas):
...
+        img_metas_list = [img_metas for _ in range(num_dec_layers)]
+
+        losses_cls, losses_bbox, losses_iou = multi_apply(
+            self.loss_single, all_cls_scores, all_bbox_preds,
+            all_gt_bboxes_list, all_gt_labels_list, img_metas_list,
+            all_gt_bboxes_ignore_list)
+
+        loss_dict = dict()
+        # loss from the last decoder layer
+        loss_dict['loss_cls'] = losses_cls[-1]
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:310]: Is it necessary to name the loss for last prediction as `loss_cls`, `loss_bbox`, and `loss_iou`?

File: mmdet/models/dense_heads/transformer_head.py [hunk 16]
```diff
@@ -245,25 +248,339 @@ def forward_single(self, x, img_metas):
...
+                shape (num_gts,).
+            gt_bboxes_ignore (Tensor): Ground truth bboxes to be
+                ignored, shape (num_ignored_gts, 4).
+            proposal_cfg (mmcv.Config): Test / postprocessing configuration,
+                if None, test_cfg would be used.
+
+        Returns:
+            dict[str, Tensor]: A dictionary of loss components.
+        """
+        assert proposal_cfg is None, 'Only when case that proposal_cfg is ' \
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:558]:
```python
assert proposal_cfg is None, '"proposal_cfg" must be None'.
```

File: mmdet/models/dense_heads/transformer_head.py (hunk 15 above)
Comment (v-qjqs) [reply mmdet/models/dense_heads/transformer_head.py:310]: Hi, currently the loss name for last predicition is consistent with the official DETR repo.
